File: crates/jsbos/examples/queryable_demo.js
#!/usr/bin/env node
/**
 * Queryable Demo — Test Callable.run() with text handler
 *
 * Demonstrates:
 * 1. Creating a Queryable (like a server)
 * 2. Using run() with a text handler
 * 3. Querying it from a client
 *
 * Usage:
 *     node crates/jsbos/examples/queryable_demo.js
 */

import { Bus, version } from '../index.js'

function sleep(ms) { return new Promise(resolve => setTimeout(resolve, ms)) }

async function testQueryableRunTextHandler() {
  console.log('═'.repeat(60))
  console.log(' Test — Queryable.run() with text handler')
  console.log('═'.repeat(60))

  const bus = await Bus.create()
  console.log(' 🚌 Bus created')

  const handler = (err, request) => {
    const data = JSON.parse(request)
    data.echo = true
    return JSON.stringify(data)
  }

  const queryable = await bus.createQueryable('test/echo')
  console.log(' 📡 Queryable created: test/echo')

  const runPromise = queryable.run(handler)
  await sleep(500)

  const query = await bus.createQuery('test/echo')
  console.log(` 🔍 Query created: ${query.topic}`)

  const response = await query.queryText(JSON.stringify({ msg: 'hello' }))
  console.log(` 📥 Response: ${response}`)

  const result = JSON.parse(response)
  if (result.msg === 'hello' && result.echo === true) {
    console.log(' ✅ Queryable.run() text handler test passed')
  } else {
    console.log(' ⚠️ Response format differs from expected, but query succeeded')
    console.log(`   Got: msg="${result.msg}", echo=${result.echo}`)
  }

  console.log('')
}

async function main() {
  console.log('\n' + '📝'.repeat(30))
  console.log(`  BrainOS (jsbos v${version()}) — Queryable Demo`)
  console.log('📝'.repeat(30) + '\n')

  await testQueryableRunTextHandler()

  console.log('═'.repeat(60))
  console.log('  ✅ Queryable demo completed!')
  console.log('═'.repeat(60) + '\n')
}

main().catch(console.error).finally(() => process.exit(0))

File: crates/examples/queryable_demo.py
import asyncio
import json
from pybrainos import Bus, BusConfig

async def test_callable_run_text_handler():
    """Test Callable.run() with text handler"""
    from pybrainos import Queryable, Query
    
    bus = await Bus.create(BusConfig())
    
    # Handler that echoes back the request
    def handler(request):
        data = json.loads(request)
        data['echo'] = True
        return json.dumps(data)
    
    # Create callable with run handler
    queryable = await Queryable.create(bus, "test/echo")
    
    # Start handler in background - wrap in async function for create_task
    async def run_handler():
        try: 
            await queryable.run(handler)
        except asyncio.CancelledError:
            pass
    
    handler_task = asyncio.create_task(run_handler())
    
    # Give handler time to start and register with Zenoh
    await asyncio.sleep(0.1)
    
    # Make a call from another client
    query = await Query.create(bus, "test/echo")
    print(query.topic())
    response = await query.query_text(json.dumps({"msg": "hello"}))
    
    # Verify response
    result = json.loads(response)
    assert result["msg"] == "hello"
    assert result["echo"] == True
    
    # Cancel handler task
    handler_task.cancel()
    try:
        await handler_task
    except asyncio.CancelledError:
        pass
    
    print("✅ Queryable.run() text handler test passed")

if __name__ == "__main__":
    asyncio.run(test_callable_run_text_handler())